### ISSUE TITLE
BigDecimal regression - to_d inconsistent with other numeric classes

### ISSUE THREAD

**Issue by jhawthorn** [ruby/bigdecimal#51]
I originally filed this on [bugs.ruby-lang.org](https://bugs.ruby-lang.org/issues/13062), not sure if this is a more appropriate place.

As of ruby 2.4.0-rc1, `BigDecimal()` was changed to raise exceptions on invalid input, which is more consistent with the other numeric types. Unfortunately, `String#to_d` now also raises errors, which is inconsistent with the other `to_*` methods (`to_i`, `to_f`), which return 0 on error.

Under ruby 2.4.0-rc1:

``` ruby
> require 'bigdecimal'
> require 'bigdecimal/util'
> "invalid".to_d
ArgumentError: invalid value for BigDecimal(): "invalid"
> "invalid".to_i
=> 0
> "invalid".to_f
=> 0.0
```

Under ruby 2.3.3 or 2.4.0preview3:

``` ruby
> "invalid".to_d
=> #<BigDecimal:55871ca1f808,'0.0',9(9)>
> "invalid".to_i
=> 0
> "invalid".to_f
=> 0.0
```

There's also a further problem that BigDecimal() still doesn't behave the same as Integer() when given a string with the number at the start:

Under ruby 2.4.0-rc1:

``` ruby
> BigDecimal("2 turtle doves")
=> 0.2e1
> Integer("2 turtle doves")
ArgumentError: invalid value for Integer(): "2 turtle doves"
> Float("2 turtle doves")
ArgumentError: invalid value for Float(): "2 turtle doves"
```

So `BigDecimal` is still inconsistent.

**Comment by mrkn** [comment]
The patch posted by Piotr Szmielew in Redmine is good to fix this.
I'll apply this patch later.
Thanks for reporting the bug.

By the way,

> I originally filed this on bugs.ruby-lang.org, not sure if this is a more appropriate place.

Both places are appropriate.
But I prefer to use GitHub issue because it is more useful than Redline.

**Comment by esse** [comment]
@mrkn Should I make pull request with that? Or is patch file provided through redmine enough?

**Comment by mrkn** [comment]
@esse Would you please submit your pull request?
I want to import your contribution as your commit.

**Comment by esse** [comment]
@mrkn pull request added: https://github.com/ruby/bigdecimal/issues/55

**Comment by mrkn** [comment]
@esse Thank you.  I'll check it at this weekend.

**Comment by abezzub** [comment]
I still see this in ruby 2.4.1. Did it not make the cut?

**Comment by mrkn** [comment]
@abezzub Could you please try `gem update bigdecimal`?

**Comment by abezzub** [comment]
I did, thanks. I see that issue is fixed in version 1.3.2. If I understand correctly, bigdecimal is shipped with Ruby, so it would be nice if a version with the fix were included in latest Ruby release.

**Comment by mrkn** [comment]
@abezzub I believe the latest bigdecimal will be included in the next update of Ruby.
BTW, now bigdecimal is a default gem, so the version of bigdecimal is maintained independently.
You can update default gems independently of Ruby version.
